agent/consul/cluster_endpoint.go

```diff
@@ -0,0 +1,298 @@
+	if c.srv.config.ConnectEnabled && c.srv.config.AutoEncryptAllowTLS {
+		conf.AutoEncrypt = &config.AutoEncrypt{TLS: true}
+	} else {
+		conf.AutoEncrypt = &config.AutoEncrypt{TLS: false}
+	}
+
+	return nil
+}
+
+// updateACLtokensInConfig will configure all of the agents ACL settings and will populate
+// the configuration with an agent token usable for all default agent operations.
+func (c *Cluster) updateACLsInConfig(opts AutoConfigOptions, conf *config.Config) error {
+	acl := &config.ACL{
+		Enabled:             c.srv.config.ACLsEnabled,
+		PolicyTTL:           c.srv.config.ACLPolicyTTL.String(),
+		RoleTTL:             c.srv.config.ACLRoleTTL.String(),
+		TokenTTL:            c.srv.config.ACLTokenTTL.String(),
+		DisabledTTL:         c.srv.config.ACLDisabledTTL.String(),
+		DownPolicy:          c.srv.config.ACLDownPolicy,
+		DefaultPolicy:       c.srv.config.ACLDefaultPolicy,
+		EnableKeyListPolicy: c.srv.config.ACLEnableKeyListPolicy,
+	}
+
+	// when ACLs are enabled we want to create a local token with a node identity
+	if c.srv.config.ACLsEnabled {
+		// we have to require local tokens or else it would require having these servers use a token with acl:write to make a
+		// token create RPC to the servers in the primary DC.
+		if !c.srv.LocalTokensEnabled() {
+			return fmt.Errorf("Agent Auto Configuration requires local token usage to be enabled in this datacenter: %s", c.srv.config.Datacenter)
+		}
+
+		// generate the accessor id
+		accessor, err := lib.GenerateUUID(c.srv.checkTokenUUID)
+		if err != nil {
+			return err
+		}
+		// generate the secret id
+		secret, err := lib.GenerateUUID(c.srv.checkTokenUUID)
+		if err != nil {
+			return err
+		}
+
+		// set up the token
+		token := structs.ACLToken{
+			AccessorID:  accessor,
+			SecretID:    secret,
+			Description: fmt.Sprintf("Auto Config Token for Node %q", opts.NodeName),
+			CreateTime:  time.Now(),
+			Local:       true,
+			NodeIdentities: []*structs.ACLNodeIdentity{
+				{
+					NodeName:   opts.NodeName,
+					Datacenter: c.srv.config.Datacenter,
+				},
+			},
+			EnterpriseMeta: *structs.DefaultEnterpriseMeta(),
+		}
+
+		req := structs.ACLTokenBatchSetRequest{
+			Tokens: structs.ACLTokens{&token},
+			CAS:    false,
+		}
+
+		// perform the request to mint the new token
+		if _, err := c.srv.raftApplyMsgpack(structs.ACLTokenSetRequestType, &req); err != nil {
+			return err
+		}
+
+		acl.Tokens = &config.ACLTokens{Agent: secret}
+	}
+
+	conf.ACL = acl
+	return nil
+}
+
+// updateJoinAddressesInConfig determines the correct gossip endpoints that clients should
+// be connecting to for joining the cluster based on the segment given in the opts parameter.
+func (c *Cluster) updateJoinAddressesInConfig(opts AutoConfigOptions, conf *config.Config) error {
+	members, err := c.srv.LANSegmentMembers(opts.SegmentName)
+	if err != nil {
+		return err
+	}
+
+	var joinAddrs []string
+	for _, m := range members {
+		if ok, _ := metadata.IsConsulServer(m); ok {
```

Comment:
Why are only servers being considered? I imagine it could help with load if a set of random nodes is being picked.

Reply:
We could consider clients and a random set but I see clients as somewhat more ephemeral than servers. While you cannot get rid of half of the servers at the same time without an outage you could get rid of all the clients. Thus when the new client goes to do the join it would be possible that those other clients are already gone. 

One other alternative would be to send the list as some number of random clients followed by all the servers. The retry join addresses should be processed in order so if those clients happened to go away it would "fallback" to the servers.

Reply:
Maybe in the future we can consider this. For now joining the cluster should be infrequent enough not to cause too much load on the servers. Also its more typical for clients to always join server addresses instead of other clients.

Reply:
👍